lxd/device/proxy.go

```diff
@@ -0,0 +1,456 @@
+		"mode":           shared.IsOctalFileMode,
+		"nat":            shared.IsBool,
+		"gid":            shared.IsUnixUserID,
+		"uid":            shared.IsUnixUserID,
+		"security.uid":   shared.IsUnixUserID,
+		"security.gid":   shared.IsUnixUserID,
+		"proxy_protocol": shared.IsBool,
+	}
+
+	err := config.ValidateDevice(rules, d.config)
+	if err != nil {
+		return err
+	}
+
+	listenAddr, err := ProxyParseAddr(d.config["listen"])
+	if err != nil {
+		return err
+	}
+
+	connectAddr, err := ProxyParseAddr(d.config["connect"])
+	if err != nil {
+		return err
+	}
+
+	if len(connectAddr.Addr) > len(listenAddr.Addr) {
+		// Cannot support single port -> multiple port
+		return fmt.Errorf("Cannot map a single port to multiple ports")
+	}
+
+	if shared.IsTrue(d.config["proxy_protocol"]) && !strings.HasPrefix(d.config["connect"], "tcp") {
+		return fmt.Errorf("The PROXY header can only be sent to tcp servers")
+	}
+
+	if (!strings.HasPrefix(d.config["listen"], "unix:") || strings.HasPrefix(d.config["listen"], "unix:@")) &&
+		(d.config["uid"] != "" || d.config["gid"] != "" || d.config["mode"] != "") {
+		return fmt.Errorf("Only proxy devices for non-abstract unix sockets can carry uid, gid, or mode properties")
+	}
+
+	if shared.IsTrue(d.config["nat"]) {
+		if d.config["bind"] != "" && d.config["bind"] != "host" {
+			return fmt.Errorf("Only host-bound proxies can use NAT")
+		}
+
+		// Support TCP <-> TCP and UDP <-> UDP
+		if listenAddr.ConnType == "unix" || connectAddr.ConnType == "unix" ||
+			listenAddr.ConnType != connectAddr.ConnType {
+			return fmt.Errorf("Proxying %s <-> %s is not supported when using NAT",
+				listenAddr.ConnType, connectAddr.ConnType)
+		}
+	}
+
+	return nil
+}
+
+// Start is run when the device is added to the container.
+func (d *proxy) Start() (*RunConfig, error) {
+	runConf := RunConfig{}
+
+	// Proxy devices have to be setup once the container is running.
+	runConf.PostStartHooks = []func() error{
+		func() error {
+			if shared.IsTrue(d.config["nat"]) {
+				return d.setupNAT()
+			}
+
+			proxyValues, err := d.setupProxyProcInfo()
+			if err != nil {
+				return err
+			}
+
+			devFileName := fmt.Sprintf("proxy.%s", d.name)
+			pidPath := filepath.Join(d.instance.DevicesPath(), devFileName)
+			logFileName := fmt.Sprintf("proxy.%s.log", d.name)
+			logPath := filepath.Join(d.instance.LogPath(), logFileName)
+
+			_, err = shared.RunCommand(
+				d.state.OS.ExecPath,
+				"forkproxy",
+				proxyValues.listenPid,
+				proxyValues.listenAddr,
+				proxyValues.connectPid,
+				proxyValues.connectAddr,
+				logPath,
+				pidPath,
+				proxyValues.listenAddrGID,
+				proxyValues.listenAddrUID,
+				proxyValues.listenAddrMode,
+				proxyValues.securityGID,
+				proxyValues.securityUID,
+				proxyValues.proxyProtocol,
+			)
+			if err != nil {
+				return fmt.Errorf("Error occurred when starting proxy device: %s", err)
+			}
+
+			// Poll log file a few times until we see "Started" to indicate successful start.
```

Comment:
@stgraber this is my solution to detecting an already used listen port and failing. It works, in that it checks the forkproxy process has actually started properly, and for any other reason will show an error to the user. Its not that nice as it depends on parsing log files though.